Quest v0.2 / G6PD : App crashes when  G6PD Test Questionnaire is completed  and submitted with a photo attachment

**Describe the bug**
App crashes when  G6PD Test Questionnaire is completed  and submitted with a photo attachment

**To Reproduce**
Steps to reproduce the behavior:
1. Go to the G6PD test 
2. Complete the survey and add a photo/replace photo
3. Submit a completed test 
4. See error

**Expected behavior**
1. Complete questionnaire with photo and submit with no crash 
2. User should be able to update photo before submission and post with no crash 

**Screenshots**
See video  - TBD

**Smartphone (please complete the following information):**
 - Device: Pixel 4a 
 - Android version: 12
 - Build number: SQ1A.220105.002

**Additional context**
- Photo takes a few seconds before showing up in its placeholder, similar when the photo is updated



> @f-odhiambo Is this a real device? Does this happens in emulator too? with the same device and API version

Error log:
```
2022-01-24 11:09:00.787 8977-8977/org.smartregister.fhircore.quest E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.smartregister.fhircore.quest, PID: 8977
    java.lang.RuntimeException: Unable to start activity ComponentInfo{org.smartregister.fhircore.quest/org.smartregister.fhircore.quest.ui.patient.details.QuestPatientDetailActivity}: kotlin.UninitializedPropertyAccessException: lateinit property appId has not been initialized
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2951)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3016)
        at android.app.ActivityThread.-wrap12(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1716)
        at android.os.Handler.dispatchMessage(Handler.java:110)
        at android.os.Looper.loop(Looper.java:232)
        at android.app.ActivityThread.main(ActivityThread.java:6806)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1103)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:964)
     Caused by: kotlin.UninitializedPropertyAccessException: lateinit property appId has not been initialized
        at org.smartregister.fhircore.engine.configuration.ConfigurationRegistry.getAppId(ConfigurationRegistry.kt:51)
        at org.smartregister.fhircore.engine.configuration.ConfigurationRegistry.workflowPointName(ConfigurationRegistry.kt:145)
        at org.smartregister.fhircore.quest.ui.patient.details.QuestPatientDetailActivity.onCreate(QuestPatientDetailActivity.kt:216)
        at android.app.Activity.performCreate(Activity.java:6974)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1118)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2904)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3016) 
        at android.app.ActivityThread.-wrap12(ActivityThread.java) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1716) 
        at android.os.Handler.dispatchMessage(Handler.java:110) 
        at android.os.Looper.loop(Looper.java:232) 
        at android.app.ActivityThread.main(ActivityThread.java:6806) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1103) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:964) 
2022-01-24 11:09:01.318 9006-9006/org.smartregister.fhircore.quest E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.smartregister.fhircore.quest, PID: 9006
    java.lang.RuntimeException: Unable to start activity ComponentInfo{org.smartregister.fhircore.quest/org.smartregister.fhircore.quest.ui.patient.register.PatientRegisterActivity}: kotlin.UninitializedPropertyAccessException: lateinit property appId has not been initialized
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2951)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3016)
        at android.app.ActivityThread.-wrap12(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1716)
        at android.os.Handler.dispatchMessage(Handler.java:110)
        at android.os.Looper.loop(Looper.java:232)
        at android.app.ActivityThread.main(ActivityThread.java:6806)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1103)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:964)
     Caused by: kotlin.UninitializedPropertyAccessException: lateinit property appId has not been initialized
        at org.smartregister.fhircore.engine.configuration.ConfigurationRegistry.getAppId(ConfigurationRegistry.kt:51)
        at org.smartregister.fhircore.engine.configuration.ConfigurationRegistry.workflowPointName(ConfigurationRegistry.kt:145)
        at org.smartregister.fhircore.engine.ui.register.RegisterViewModel.<init>(RegisterViewModel.kt:153)
        at org.smartregister.fhircore.quest.DaggerQuestApplication_HiltComponents_SingletonC$ViewModelCImpl.registerViewModel(DaggerQuestApplication_HiltComponents_SingletonC.java:755)
        at org.smartregister.fhircore.quest.DaggerQuestApplication_HiltComponents_SingletonC$ViewModelCImpl.access$3000(DaggerQuestApplication_HiltComponents_SingletonC.java:700)
        at org.smartregister.fhircore.quest.DaggerQuestApplication_HiltComponents_SingletonC$ViewModelCImpl$SwitchingProvider.get(DaggerQuestApplication_HiltComponents_SingletonC.java:816)
        at dagger.hilt.android.internal.lifecycle.HiltViewModelFactory$1.create(HiltViewModelFactory.java:100)
        at androidx.lifecycle.AbstractSavedStateViewModelFactory.create(AbstractSavedStateViewModelFactory.java:69)
        at androidx.lifecycle.AbstractSavedStateViewModelFactory.create(AbstractSavedStateViewModelFactory.java:84)
        at dagger.hilt.android.internal.lifecycle.HiltViewModelFactory.create(HiltViewModelFactory.java:109)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:187)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:150)
        at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:54)
        at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:41)
        at org.smartregister.fhircore.engine.ui.register.BaseRegisterActivity.getRegisterViewModel(BaseRegisterActivity.kt:105)
        at org.smartregister.fhircore.engine.ui.register.BaseRegisterActivity.runSync(BaseRegisterActivity.kt:430)
        at org.smartregister.fhircore.engine.sync.SyncBroadcaster.registerSyncInitiator(SyncHandlers.kt:56)
        at org.smartregister.fhircore.engine.ui.register.BaseRegisterActivity.onCreate(BaseRegisterActivity.kt:129)
        at org.smartregister.fhircore.quest.ui.patient.register.PatientRegisterActivity.onCreate(PatientRegisterActivity.kt:40)
        at android.app.Activity.performCreate(Activity.java:6974)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1118)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2904)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3016) 
        at android.app.ActivityThread.-wrap12(ActivityThread.java) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1716) 
        at android.os.Handler.dispatchMessage(Handler.java:110) 
        at android.os.Looper.loop(Looper.java:232) 
        at android.app.ActivityThread.main(ActivityThread.java:6806) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1103) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:964)
```

Out of memory
```
2022-01-25 15:18:24.455 24527-24527/org.smartregister.fhircore.quest E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.smartregister.fhircore.quest, PID: 24527
    java.lang.RuntimeException: android.os.TransactionTooLargeException: data parcel size 2794020 bytes
        at android.app.ActivityClient.finishActivity(ActivityClient.java:156)
        at android.app.Activity.finish(Activity.java:6465)
        at android.app.Activity.finish(Activity.java:6499)
        at org.smartregister.fhircore.engine.ui.questionnaire.QuestionnaireActivity.finishActivity(QuestionnaireActivity.kt:300)
        at org.smartregister.fhircore.engine.ui.questionnaire.QuestionnaireActivity.postSaveSuccessful(QuestionnaireActivity.kt:289)
        at org.smartregister.fhircore.engine.ui.questionnaire.QuestionnaireActivity.handleQuestionnaireSubmit$lambda-4(QuestionnaireActivity.kt:267)
        at org.smartregister.fhircore.engine.ui.questionnaire.QuestionnaireActivity.$r8$lambda$Q3ZvH3-oznfR02Y6M-FSEGTDJaU(Unknown Source:0)
        at org.smartregister.fhircore.engine.ui.questionnaire.QuestionnaireActivity$$ExternalSyntheticLambda0.onChanged(Unknown Source:6)
        at androidx.lifecycle.LiveData.considerNotify(LiveData.java:133)
        at androidx.lifecycle.LiveData.dispatchingValue(LiveData.java:151)
        at androidx.lifecycle.LiveData.setValue(LiveData.java:309)
        at androidx.lifecycle.MutableLiveData.setValue(MutableLiveData.java:50)
        at androidx.lifecycle.LiveData$1.run(LiveData.java:93)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
     Caused by: android.os.TransactionTooLargeException: data parcel size 2794020 bytes
        at android.os.BinderProxy.transactNative(Native Method)
        at android.os.BinderProxy.transact(BinderProxy.java:571)
        at android.app.IActivityClientController$Stub$Proxy.finishActivity(IActivityClientController.java:1496)
        at android.app.ActivityClient.finishActivity(ActivityClient.java:153)
        at android.app.Activity.finish(Activity.java:6465) 
        at android.app.Activity.finish(Activity.java:6499) 
        at org.smartregister.fhircore.engine.ui.questionnaire.QuestionnaireActivity.finishActivity(QuestionnaireActivity.kt:300) 
        at org.smartregister.fhircore.engine.ui.questionnaire.QuestionnaireActivity.postSaveSuccessful(QuestionnaireActivity.kt:289) 
        at org.smartregister.fhircore.engine.ui.questionnaire.QuestionnaireActivity.handleQuestionnaireSubmit$lambda-4(QuestionnaireActivity.kt:267) 
        at org.smartregister.fhircore.engine.ui.questionnaire.QuestionnaireActivity.$r8$lambda$Q3ZvH3-oznfR02Y6M-FSEGTDJaU(Unknown Source:0) 
        at org.smartregister.fhircore.engine.ui.questionnaire.QuestionnaireActivity$$ExternalSyntheticLambda0.onChanged(Unknown Source:6) 
        at androidx.lifecycle.LiveData.considerNotify(LiveData.java:133) 
        at androidx.lifecycle.LiveData.dispatchingValue(LiveData.java:151) 
        at androidx.lifecycle.LiveData.setValue(LiveData.java:309) 
        at androidx.lifecycle.MutableLiveData.setValue(MutableLiveData.java:50) 
        at androidx.lifecycle.LiveData$1.run(LiveData.java:93) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loopOnce(Looper.java:201) 
        at android.os.Looper.loop(Looper.java:288) 
        at android.app.ActivityThread.main(ActivityThread.java:7839) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003) 
```
CC @maimoonak @FikriMilano

--- a/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
+++ b/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
@@ -26,7 +26,6 @@ import androidx.activity.result.contract.ActivityResultContracts
 import androidx.fragment.app.Fragment
 import androidx.lifecycle.LifecycleCoroutineScope
 import androidx.lifecycle.lifecycleScope
-import androidx.lifecycle.withResumed
 import com.bumptech.glide.Glide
 import com.bumptech.glide.load.DecodeFormat
 import com.bumptech.glide.request.RequestOptions
@@ -38,7 +37,6 @@ import com.google.android.fhir.datacapture.views.QuestionnaireItemViewItem
 import com.google.android.material.button.MaterialButton
 import kotlinx.coroutines.launch
 import org.hl7.fhir.r4.model.Attachment
-import org.hl7.fhir.r4.model.QuestionnaireResponse
 import org.hl7.fhir.r4.model.QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent
 import org.smartregister.fhircore.engine.R
 import org.smartregister.fhircore.engine.databinding.CustomPhotoCaptureLayoutBinding
@@ -72,12 +70,12 @@ class CustomPhotoCaptureFactory(
   internal fun registerCameraLauncher(): ActivityResultLauncher<Void> {
     return fragment.registerForActivityResult(ActivityResultContracts.TakePicturePreview()) { bitmap
       ->
-        if (bitmap != null) {
-          loadThumbnail(bitmap)
-          val bytes = bitmap.encodeToByteArray()
-          populateQuestionnaireResponse(bytes)
-          onAnswerChanged.invoke()
-        }
+      if (bitmap != null) {
+        loadThumbnail(bitmap)
+        val bytes = bitmap.encodeToByteArray()
+        populateQuestionnaireResponse(bytes)
+        onAnswerChanged.invoke()
+      }
     }
   }
 
@@ -98,7 +96,8 @@ class CustomPhotoCaptureFactory(
 
   internal fun populateQuestionnaireResponse(imageBytes: ByteArray) {
     answers.clear()
-    val answer = QuestionnaireResponseItemAnswerComponent().apply {
+    val answer =
+      QuestionnaireResponseItemAnswerComponent().apply {
         value =
           Attachment().apply {
             contentType = CONTENT_TYPE
@@ -150,16 +149,16 @@ class CustomPhotoCaptureFactory(
       }
 
       override fun displayValidationResult(validationResult: ValidationResult) {
-            tvError.text =
-                    if (validationResult.getSingleStringValidationMessage() == "") null
-                    else validationResult.getSingleStringValidationMessage()
+        tvError.text =
+          if (validationResult.getSingleStringValidationMessage() == "") null
+          else validationResult.getSingleStringValidationMessage()
       }
 
       override fun setReadOnly(isReadOnly: Boolean) {
         ivThumbnail.isEnabled = !isReadOnly
         btnTakePhoto.apply {
-        isEnabled = !isReadOnly
-        alpha = if (isReadOnly) 0.6F else 1F
+          isEnabled = !isReadOnly
+          alpha = if (isReadOnly) 0.6F else 1F
         }
       }
     }

--- a/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
+++ b/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
@@ -18,9 +18,6 @@ package org.smartregister.fhircore.engine.ui.questionnaire
 
 import android.content.Context
 import android.graphics.Bitmap
-import android.graphics.BitmapFactory
-import android.net.Uri
-import android.os.Environment
 import android.view.View
 import android.widget.ImageView
 import android.widget.TextView
@@ -30,18 +27,17 @@ import androidx.fragment.app.Fragment
 import androidx.lifecycle.LifecycleCoroutineScope
 import androidx.lifecycle.lifecycleScope
 import com.bumptech.glide.Glide
+import com.bumptech.glide.load.DecodeFormat
+import com.bumptech.glide.request.RequestOptions
 import com.google.android.fhir.datacapture.validation.ValidationResult
+import com.google.android.fhir.datacapture.validation.getSingleStringValidationMessage
 import com.google.android.fhir.datacapture.views.QuestionnaireItemViewHolderDelegate
 import com.google.android.fhir.datacapture.views.QuestionnaireItemViewHolderFactory
 import com.google.android.fhir.datacapture.views.QuestionnaireItemViewItem
 import com.google.android.material.button.MaterialButton
-import id.zelory.compressor.Compressor.compress
-import id.zelory.compressor.constraint.quality
-import id.zelory.compressor.constraint.size
-import id.zelory.compressor.extension
-import java.io.File
 import kotlinx.coroutines.launch
 import org.hl7.fhir.r4.model.Attachment
+import org.hl7.fhir.r4.model.QuestionnaireResponse
 import org.hl7.fhir.r4.model.QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent
 import org.smartregister.fhircore.engine.R
 import org.smartregister.fhircore.engine.databinding.CustomPhotoCaptureLayoutBinding
@@ -51,7 +47,6 @@ import org.smartregister.fhircore.engine.util.extension.decodeToBitmap
 import org.smartregister.fhircore.engine.util.extension.encodeToByteArray
 import org.smartregister.fhircore.engine.util.extension.hide
 import org.smartregister.fhircore.engine.util.extension.show
-import org.smartregister.fhircore.engine.util.extension.toUri
 
 class CustomPhotoCaptureFactory(
   val fragment: Fragment,
@@ -63,32 +58,26 @@ class CustomPhotoCaptureFactory(
   lateinit var tvHeader: TextView
   lateinit var ivThumbnail: ImageView
   lateinit var btnTakePhoto: MaterialButton
-  lateinit var imageFile: File
+  lateinit var errorTextView: TextView
   var context: Context = fragment.requireContext()
-  var cameraLauncher: ActivityResultLauncher<Uri>
-  var questionnaireResponse: QuestionnaireResponseItemAnswerComponent
+  var cameraLauncher: ActivityResultLauncher<Void>
+  var questionnaireResponse: QuestionnaireResponse.QuestionnaireResponseItemComponent
+  var answers: MutableList<QuestionnaireResponseItemAnswerComponent> = mutableListOf()
+  lateinit var onAnswerChanged: () -> Unit
 
   init {
     cameraLauncher = registerCameraLauncher()
-    questionnaireResponse = QuestionnaireResponseItemAnswerComponent()
+    questionnaireResponse = QuestionnaireResponse.QuestionnaireResponseItemComponent()
   }
 
-  internal fun registerCameraLauncher(): ActivityResultLauncher<Uri> {
-    return fragment.registerForActivityResult(ActivityResultContracts.TakePicture()) { result ->
-      if (result) {
-        lifecycleScope
-          .launch(dispatcher.io()) {
-            imageFile =
-              compress(context, imageFile) {
-                quality(30)
-                size(64_000)
-              }
-            val imageBitmap = BitmapFactory.decodeFile(imageFile.absolutePath)
-            loadThumbnail(imageBitmap)
-            val imageBytes = imageBitmap.encodeToByteArray()
-            populateQuestionnaireResponse(imageBytes)
-          }
-          .invokeOnCompletion { imageFile.delete() }
+  internal fun registerCameraLauncher(): ActivityResultLauncher<Void> {
+    return fragment.registerForActivityResult(ActivityResultContracts.TakePicturePreview()) { bitmap
+      ->
+      lifecycleScope.launch {
+        loadThumbnail(bitmap)
+        val bytes = bitmap.encodeToByteArray()
+        populateQuestionnaireResponse(bytes)
+        onAnswerChanged.invoke()
       }
     }
   }
@@ -98,7 +87,10 @@ class CustomPhotoCaptureFactory(
   ) {
     lifecycleScope.launch(dispatcher.main()) {
       ivThumbnail.apply {
-        Glide.with(fragment).load(imageBitmap).into(this)
+        Glide.with(fragment)
+          .load(imageBitmap)
+          .apply(RequestOptions().format(DecodeFormat.PREFER_ARGB_8888))
+          .into(this)
         show()
       }
       btnTakePhoto.text = fragment.getString(R.string.replace_photo)
@@ -106,23 +98,20 @@ class CustomPhotoCaptureFactory(
   }
 
   internal fun populateQuestionnaireResponse(imageBytes: ByteArray) {
-    questionnaireResponse.value =
-      Attachment().apply {
-        contentType = CONTENT_TYPE
-        data = imageBytes
+    answers.clear()
+    val answer =
+      QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
+        value =
+          Attachment().apply {
+            contentType = CONTENT_TYPE
+            data = imageBytes
+          }
       }
-  }
-
-  internal fun createImageFile(): File {
-    return File.createTempFile(
-      PREFIX_BITMAP,
-      ".${Bitmap.CompressFormat.JPEG.extension()}",
-      context.getExternalFilesDir(Environment.DIRECTORY_PICTURES)
-    )
+    answers.add(answer)
   }
 
   internal fun launchCamera() {
-    imageFile.toUri(context, AUTHORITY_FILE_PROVIDER).also { uri -> cameraLauncher.launch(uri) }
+    cameraLauncher.launch(null)
   }
 
   override fun getQuestionnaireItemViewHolderDelegate(): QuestionnaireItemViewHolderDelegate =
@@ -136,7 +125,9 @@ class CustomPhotoCaptureFactory(
           tvHeader = binding.tvHeader
           ivThumbnail = binding.ivThumbnail
           btnTakePhoto = binding.btnTakePhoto
+          errorTextView = binding.errorTextView as TextView
         }
+        onAnswerChanged = { onAnswerChanged(context) }
       }
 
       override fun bind(questionnaireItemViewItem: QuestionnaireItemViewItem) {
@@ -149,22 +140,18 @@ class CustomPhotoCaptureFactory(
           tvPrefix.hide(true)
         }
         tvHeader.text = questionnaireItemViewItem.questionnaireItem.text
-        btnTakePhoto.setOnClickListener {
-          imageFile = createImageFile()
-          launchCamera()
-        }
+        btnTakePhoto.setOnClickListener { launchCamera() }
         questionnaireItemViewItem.singleAnswerOrNull?.valueAttachment?.data?.decodeToBitmap()
           ?.let { imageBitmap -> loadThumbnail(imageBitmap) }
-        setReadOnly(questionnaireItemViewItem.questionnaireItem.readOnly)
-        questionnaireItemViewItem.singleAnswerOrNull = questionnaireResponse
+
+        if (!questionnaireItemViewItem.questionnaireItem.readOnly) {
+          questionnaireItemViewItem.questionnaireResponseItem.answer = answers
+        }
       }
 
       override fun displayValidationResult(validationResult: ValidationResult) {
         // Custom validation message
       }
-
-      // TODO -> Should use the overridden setReadOnly()
-      // after upgrading Data Capture library to Beta
 
       override fun setReadOnly(isReadOnly: Boolean) {
         ivThumbnail.isEnabled = !isReadOnly

--- a/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
+++ b/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
@@ -26,6 +26,7 @@ import androidx.activity.result.contract.ActivityResultContracts
 import androidx.fragment.app.Fragment
 import androidx.lifecycle.LifecycleCoroutineScope
 import androidx.lifecycle.lifecycleScope
+import androidx.lifecycle.withResumed
 import com.bumptech.glide.Glide
 import com.bumptech.glide.load.DecodeFormat
 import com.bumptech.glide.request.RequestOptions
@@ -58,27 +59,25 @@ class CustomPhotoCaptureFactory(
   lateinit var tvHeader: TextView
   lateinit var ivThumbnail: ImageView
   lateinit var btnTakePhoto: MaterialButton
-  lateinit var errorTextView: TextView
+  lateinit var tvError: TextView
   var context: Context = fragment.requireContext()
   var cameraLauncher: ActivityResultLauncher<Void>
-  var questionnaireResponse: QuestionnaireResponse.QuestionnaireResponseItemComponent
   var answers: MutableList<QuestionnaireResponseItemAnswerComponent> = mutableListOf()
   lateinit var onAnswerChanged: () -> Unit
 
   init {
     cameraLauncher = registerCameraLauncher()
-    questionnaireResponse = QuestionnaireResponse.QuestionnaireResponseItemComponent()
   }
 
   internal fun registerCameraLauncher(): ActivityResultLauncher<Void> {
     return fragment.registerForActivityResult(ActivityResultContracts.TakePicturePreview()) { bitmap
       ->
-      lifecycleScope.launch {
-        loadThumbnail(bitmap)
-        val bytes = bitmap.encodeToByteArray()
-        populateQuestionnaireResponse(bytes)
-        onAnswerChanged.invoke()
-      }
+        if (bitmap != null) {
+          loadThumbnail(bitmap)
+          val bytes = bitmap.encodeToByteArray()
+          populateQuestionnaireResponse(bytes)
+          onAnswerChanged.invoke()
+        }
     }
   }
 
@@ -99,8 +98,7 @@ class CustomPhotoCaptureFactory(
 
   internal fun populateQuestionnaireResponse(imageBytes: ByteArray) {
     answers.clear()
-    val answer =
-      QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
+    val answer = QuestionnaireResponseItemAnswerComponent().apply {
         value =
           Attachment().apply {
             contentType = CONTENT_TYPE
@@ -125,7 +123,7 @@ class CustomPhotoCaptureFactory(
           tvHeader = binding.tvHeader
           ivThumbnail = binding.ivThumbnail
           btnTakePhoto = binding.btnTakePhoto
-          errorTextView = binding.errorTextView as TextView
+          tvError = binding.tvError as TextView
         }
         onAnswerChanged = { onAnswerChanged(context) }
       }
@@ -141,31 +139,32 @@ class CustomPhotoCaptureFactory(
         }
         tvHeader.text = questionnaireItemViewItem.questionnaireItem.text
         btnTakePhoto.setOnClickListener { launchCamera() }
-        questionnaireItemViewItem.singleAnswerOrNull?.valueAttachment?.data?.decodeToBitmap()
-          ?.let { imageBitmap -> loadThumbnail(imageBitmap) }
-
+        questionnaireItemViewItem.singleAnswerOrNull?.valueAttachment?.let { attachment ->
+          loadThumbnail(attachment.data.decodeToBitmap())
+          answers.clear()
+          answers.add(QuestionnaireResponseItemAnswerComponent().apply { value = attachment })
+        }
         if (!questionnaireItemViewItem.questionnaireItem.readOnly) {
           questionnaireItemViewItem.questionnaireResponseItem.answer = answers
         }
       }
 
       override fun displayValidationResult(validationResult: ValidationResult) {
-        lifecycleScope.launchWhenResumed {
-          errorTextView.text =
-            if (validationResult.getSingleStringValidationMessage() == "") null
-            else validationResult.getSingleStringValidationMessage()
-        }
+            tvError.text =
+                    if (validationResult.getSingleStringValidationMessage() == "") null
+                    else validationResult.getSingleStringValidationMessage()
       }
 
       override fun setReadOnly(isReadOnly: Boolean) {
         ivThumbnail.isEnabled = !isReadOnly
-        btnTakePhoto.isEnabled = !isReadOnly
+        btnTakePhoto.apply {
+        isEnabled = !isReadOnly
+        alpha = if (isReadOnly) 0.6F else 1F
+        }
       }
     }
 
   companion object {
-    private const val AUTHORITY_FILE_PROVIDER = "org.smartregister.fhircore.fileprovider"
     const val CONTENT_TYPE = "image/jpg"
-    private const val PREFIX_BITMAP = "BITMAP_"
   }
 }

--- a/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
+++ b/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactory.kt
@@ -150,7 +150,11 @@ class CustomPhotoCaptureFactory(
       }
 
       override fun displayValidationResult(validationResult: ValidationResult) {
-        // Custom validation message
+        lifecycleScope.launchWhenResumed {
+          errorTextView.text =
+            if (validationResult.getSingleStringValidationMessage() == "") null
+            else validationResult.getSingleStringValidationMessage()
+        }
       }
 
       override fun setReadOnly(isReadOnly: Boolean) {

--- a/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/QuestionnaireViewModel.kt
+++ b/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/QuestionnaireViewModel.kt
@@ -31,6 +31,7 @@ import java.util.Date
 import java.util.UUID
 import javax.inject.Inject
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
 import org.hl7.fhir.r4.context.IWorkerContext
@@ -193,14 +194,25 @@ constructor(
           editQuestionnaireResponse!!.deleteRelatedResources(defaultRepository)
         }
 
-        questionnaire.cqfLibraryIds().forEach {
-          val patient =
-            if (questionnaireResponse.hasSubject())
-              loadPatient(questionnaireResponse.subject.extractId())
-            else null
-          val output = libraryEvaluator.runCqlLibrary(it, patient, bundle, defaultRepository)
-          if (output.isNotEmpty()) extractionProgressMessage.postValue(output.joinToString("\n"))
-        }
+        questionnaire
+          .cqfLibraryIds()
+          .map {
+            val patient =
+              if (questionnaireResponse.hasSubject())
+                loadPatient(questionnaireResponse.subject.extractId())
+              else null
+            async(Dispatchers.Default) {
+              libraryEvaluator.runCqlLibrary(it, patient, bundle, defaultRepository)
+            }
+          }
+          .map { jobOutput ->
+            jobOutput.join()
+            jobOutput
+          }
+          .forEach { output ->
+            if (output.await().isNotEmpty())
+              extractionProgressMessage.postValue(output.await().joinToString("\n"))
+          }
       } else {
         saveQuestionnaireResponse(questionnaire, questionnaireResponse)
       }

--- a/android/engine/src/main/res/layout/custom_photo_capture_layout.xml
+++ b/android/engine/src/main/res/layout/custom_photo_capture_layout.xml
@@ -13,12 +13,14 @@
         android:orientation="horizontal">
 
         <TextView
+            style="?attr/headerTextAppearanceQuestionnaire"
             android:id="@+id/tv_prefix"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
             android:visibility="gone" />
 
         <TextView
+            style="?attr/headerTextAppearanceQuestionnaire"
             android:id="@+id/tv_header"
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
@@ -38,8 +40,8 @@
         android:visibility="gone" />
 
     <com.google.android.material.button.MaterialButton
-        android:id="@+id/btn_take_photo"
         style="@style/Widget.AppCompat.Button.Borderless"
+        android:id="@+id/btn_take_photo"
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
         android:letterSpacing="0"
@@ -49,5 +51,10 @@
         app:icon="@drawable/ic_baseline_photo_camera"
         app:iconPadding="@dimen/item_margin_horizontal"
         app:iconTint="@color/colorPrimaryLight" />
+
+    <include
+        android:id="@+id/error_text_view"
+        layout="@layout/input_error_text_view"
+        />
 
 </LinearLayout>

--- a/android/engine/src/main/res/layout/custom_photo_capture_layout.xml
+++ b/android/engine/src/main/res/layout/custom_photo_capture_layout.xml
@@ -17,15 +17,14 @@
             android:id="@+id/tv_prefix"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
+            android:paddingEnd="@dimen/prefix_padding_end"
             android:visibility="gone" />
 
         <TextView
             style="?attr/headerTextAppearanceQuestionnaire"
             android:id="@+id/tv_header"
             android:layout_width="match_parent"
-            android:layout_height="wrap_content"
-            android:layout_marginStart="@dimen/item_margin_vertical"
-            android:layout_marginEnd="@dimen/item_margin_horizontal" />
+            android:layout_height="wrap_content" />
 
     </LinearLayout>
 
@@ -53,7 +52,7 @@
         app:iconTint="@color/colorPrimaryLight" />
 
     <include
-        android:id="@+id/error_text_view"
+        android:id="@+id/tv_error"
         layout="@layout/input_error_text_view"
         />
 

--- a/android/engine/src/test/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactoryTest.kt
+++ b/android/engine/src/test/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactoryTest.kt
@@ -17,6 +17,8 @@
 package org.smartregister.fhircore.engine.ui.questionnaire
 
 import android.app.Application
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
 import android.net.Uri
 import android.view.View
 import android.widget.ImageView
@@ -24,6 +26,8 @@ import android.widget.TextView
 import androidx.activity.result.ActivityResultCallback
 import androidx.activity.result.contract.ActivityResultContract
 import androidx.test.core.app.ApplicationProvider
+import com.google.android.fhir.datacapture.validation.ValidationResult
+import com.google.android.fhir.datacapture.validation.getSingleStringValidationMessage
 import com.google.android.fhir.datacapture.views.QuestionnaireItemViewItem
 import com.google.android.material.button.MaterialButton
 import dagger.hilt.android.testing.HiltAndroidRule
@@ -33,10 +37,13 @@ import io.mockk.mockk
 import io.mockk.slot
 import io.mockk.spyk
 import io.mockk.verify
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
 import java.io.File
 import org.hl7.fhir.r4.model.Attachment
 import org.hl7.fhir.r4.model.Questionnaire
 import org.hl7.fhir.r4.model.QuestionnaireResponse
+import org.hl7.fhir.r4.model.StringType
 import org.junit.Assert
 import org.junit.Before
 import org.junit.Rule
@@ -60,24 +67,22 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
     every { fragment.requireContext() } returns context
 
     val photoCaptureFactory = spyk(CustomPhotoCaptureFactory(fragment))
-    val file = mockk<File>(relaxed = true)
-    every { file.delete() } returns true
-    every { photoCaptureFactory.imageFile } returns file
 
-    val callback = slot<ActivityResultCallback<Boolean>>()
+    every { photoCaptureFactory.onAnswerChanged } returns mockk(relaxed = true)
+
+    val callback = slot<ActivityResultCallback<Bitmap>>()
     every {
       fragment.registerForActivityResult(
-        any<ActivityResultContract<Uri, Boolean>>(),
+        any<ActivityResultContract<Void, Bitmap>>(),
         capture(callback)
       )
     } returns mockk()
 
     photoCaptureFactory.registerCameraLauncher()
-    photoCaptureFactory.loadThumbnail(mockk())
-    photoCaptureFactory.populateQuestionnaireResponse("text".encodeToByteArray())
 
-    callback.captured.onActivityResult(true)
-    callback.captured.onActivityResult(false)
+    val byteArray = "image".encodeToByteArray()
+    val bitmap = BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
+    callback.captured.onActivityResult(bitmap)
 
     verify { photoCaptureFactory.loadThumbnail(any()) }
     verify { photoCaptureFactory.populateQuestionnaireResponse(any()) }
@@ -89,16 +94,18 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
     every { fragment.requireContext() } returns context
 
     val photoCaptureFactory = spyk(CustomPhotoCaptureFactory(fragment))
-    val imageBytes = "file".encodeToByteArray()
+    val imageBytes = "image".encodeToByteArray()
+
+    Assert.assertEquals(photoCaptureFactory.answers.size, 0)
 
     photoCaptureFactory.populateQuestionnaireResponse(imageBytes)
 
-    Assert.assertTrue(photoCaptureFactory.questionnaireResponse.hasValueAttachment())
+    Assert.assertEquals(photoCaptureFactory.answers.size, 1)
     Assert.assertEquals(
       CustomPhotoCaptureFactory.CONTENT_TYPE,
-      photoCaptureFactory.questionnaireResponse.valueAttachment.contentType
+      photoCaptureFactory.answers.firstOrNull()?.valueAttachment?.contentType
     )
-    Assert.assertEquals(imageBytes, photoCaptureFactory.questionnaireResponse.valueAttachment.data)
+    Assert.assertEquals(imageBytes, photoCaptureFactory.answers.firstOrNull()?.valueAttachment?.data)
   }
 
   @Test
@@ -139,7 +146,6 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
 
     callback.captured.onClick(btnTakePhoto)
 
-    verify { photoCaptureFactory.createImageFile() }
     verify { tvPrefix.text = "1." }
     verify { tvPrefix.visibility = View.VISIBLE }
     verify { tvHeader.text = "Photo of device" }
@@ -183,8 +189,79 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
 
     callback.captured.onClick(btnTakePhoto)
 
-    verify { photoCaptureFactory.createImageFile() }
     verify { tvPrefix.visibility = View.GONE }
     verify { tvHeader.text = "Photo of device" }
+  }
+
+  @Test
+  fun testPhotoCaptureShouldDisplayValidationResult() {
+    val fragment = spyk<FhirCoreQuestionnaireFragment>()
+    every { fragment.requireContext() } returns context
+
+    val photoCaptureFactory = spyk(CustomPhotoCaptureFactory(fragment))
+
+    val tvError = mockk<TextView>(relaxed = true)
+    every { photoCaptureFactory.tvError } returns tvError
+
+    val validationResult = ValidationResult(false, listOf("Error"))
+    photoCaptureFactory.getQuestionnaireItemViewHolderDelegate().displayValidationResult(validationResult)
+
+    verify { tvError.text = "Error" }
+  }
+
+  @Test
+  fun testPhotoCaptureShouldNotDisplayValidationResult() {
+    val fragment = spyk<FhirCoreQuestionnaireFragment>()
+    every { fragment.requireContext() } returns context
+
+    val photoCaptureFactory = spyk(CustomPhotoCaptureFactory(fragment))
+
+    val tvError = mockk<TextView>(relaxed = true)
+    every { photoCaptureFactory.tvError } returns tvError
+
+    val validationResult = ValidationResult(true, listOf())
+    photoCaptureFactory.getQuestionnaireItemViewHolderDelegate().displayValidationResult(validationResult)
+
+    verify { tvError.text = null }
+  }
+
+  @Test
+  fun testPhotoCaptureShouldSetReadOnly() {
+    val fragment = spyk<FhirCoreQuestionnaireFragment>()
+    every { fragment.requireContext() } returns context
+
+    val photoCaptureFactory = spyk(CustomPhotoCaptureFactory(fragment))
+
+    val ivThumbnail = mockk<ImageView>(relaxed = true)
+    every { photoCaptureFactory.ivThumbnail } returns ivThumbnail
+
+    val btnTakePhoto = mockk<MaterialButton>(relaxed = true)
+    every { photoCaptureFactory.btnTakePhoto } returns btnTakePhoto
+
+    photoCaptureFactory.getQuestionnaireItemViewHolderDelegate().setReadOnly(true)
+
+    verify{ ivThumbnail.isEnabled = false}
+    verify{ btnTakePhoto.isEnabled = false}
+    verify{ btnTakePhoto.alpha = 0.6F}
+  }
+
+  @Test
+  fun testPhotoCaptureShouldNotSetReadOnly() {
+    val fragment = spyk<FhirCoreQuestionnaireFragment>()
+    every { fragment.requireContext() } returns context
+
+    val photoCaptureFactory = spyk(CustomPhotoCaptureFactory(fragment))
+
+    val ivThumbnail = mockk<ImageView>(relaxed = true)
+    every { photoCaptureFactory.ivThumbnail } returns ivThumbnail
+
+    val btnTakePhoto = mockk<MaterialButton>(relaxed = true)
+    every { photoCaptureFactory.btnTakePhoto } returns btnTakePhoto
+
+    photoCaptureFactory.getQuestionnaireItemViewHolderDelegate().setReadOnly(false)
+
+    verify{ ivThumbnail.isEnabled = true}
+    verify{ btnTakePhoto.isEnabled = true}
+    verify{ btnTakePhoto.alpha = 1F}
   }
 }

--- a/android/engine/src/test/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactoryTest.kt
+++ b/android/engine/src/test/java/org/smartregister/fhircore/engine/ui/questionnaire/CustomPhotoCaptureFactoryTest.kt
@@ -19,7 +19,6 @@ package org.smartregister.fhircore.engine.ui.questionnaire
 import android.app.Application
 import android.graphics.Bitmap
 import android.graphics.BitmapFactory
-import android.net.Uri
 import android.view.View
 import android.widget.ImageView
 import android.widget.TextView
@@ -27,7 +26,6 @@ import androidx.activity.result.ActivityResultCallback
 import androidx.activity.result.contract.ActivityResultContract
 import androidx.test.core.app.ApplicationProvider
 import com.google.android.fhir.datacapture.validation.ValidationResult
-import com.google.android.fhir.datacapture.validation.getSingleStringValidationMessage
 import com.google.android.fhir.datacapture.views.QuestionnaireItemViewItem
 import com.google.android.material.button.MaterialButton
 import dagger.hilt.android.testing.HiltAndroidRule
@@ -37,13 +35,9 @@ import io.mockk.mockk
 import io.mockk.slot
 import io.mockk.spyk
 import io.mockk.verify
-import kotlinx.coroutines.Job
-import kotlinx.coroutines.launch
-import java.io.File
 import org.hl7.fhir.r4.model.Attachment
 import org.hl7.fhir.r4.model.Questionnaire
 import org.hl7.fhir.r4.model.QuestionnaireResponse
-import org.hl7.fhir.r4.model.StringType
 import org.junit.Assert
 import org.junit.Before
 import org.junit.Rule
@@ -105,7 +99,10 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
       CustomPhotoCaptureFactory.CONTENT_TYPE,
       photoCaptureFactory.answers.firstOrNull()?.valueAttachment?.contentType
     )
-    Assert.assertEquals(imageBytes, photoCaptureFactory.answers.firstOrNull()?.valueAttachment?.data)
+    Assert.assertEquals(
+      imageBytes,
+      photoCaptureFactory.answers.firstOrNull()?.valueAttachment?.data
+    )
   }
 
   @Test
@@ -204,7 +201,9 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
     every { photoCaptureFactory.tvError } returns tvError
 
     val validationResult = ValidationResult(false, listOf("Error"))
-    photoCaptureFactory.getQuestionnaireItemViewHolderDelegate().displayValidationResult(validationResult)
+    photoCaptureFactory
+      .getQuestionnaireItemViewHolderDelegate()
+      .displayValidationResult(validationResult)
 
     verify { tvError.text = "Error" }
   }
@@ -220,7 +219,9 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
     every { photoCaptureFactory.tvError } returns tvError
 
     val validationResult = ValidationResult(true, listOf())
-    photoCaptureFactory.getQuestionnaireItemViewHolderDelegate().displayValidationResult(validationResult)
+    photoCaptureFactory
+      .getQuestionnaireItemViewHolderDelegate()
+      .displayValidationResult(validationResult)
 
     verify { tvError.text = null }
   }
@@ -240,9 +241,9 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
 
     photoCaptureFactory.getQuestionnaireItemViewHolderDelegate().setReadOnly(true)
 
-    verify{ ivThumbnail.isEnabled = false}
-    verify{ btnTakePhoto.isEnabled = false}
-    verify{ btnTakePhoto.alpha = 0.6F}
+    verify { ivThumbnail.isEnabled = false }
+    verify { btnTakePhoto.isEnabled = false }
+    verify { btnTakePhoto.alpha = 0.6F }
   }
 
   @Test
@@ -260,8 +261,8 @@ class CustomPhotoCaptureFactoryTest : RobolectricTest() {
 
     photoCaptureFactory.getQuestionnaireItemViewHolderDelegate().setReadOnly(false)
 
-    verify{ ivThumbnail.isEnabled = true}
-    verify{ btnTakePhoto.isEnabled = true}
-    verify{ btnTakePhoto.alpha = 1F}
+    verify { ivThumbnail.isEnabled = true }
+    verify { btnTakePhoto.isEnabled = true }
+    verify { btnTakePhoto.alpha = 1F }
   }
 }

